runtimes/js/encore.dev/internal/appinit/mod.ts

```diff
@@ -90,36 +181,87 @@ function transformHandler(h: Handler): Handler {
           stream = new IterableSocket(stream);
         }
 
-        // handshake payload
-        const payload = req.payload();
-        return payload !== null
-          ? h.handler(payload, stream)
-          : h.handler(stream);
+        if (middlewares.length === 0) {
+          const payload = req.payload();
+          return toResponse(
+            payload !== null
+              ? h.apiRoute.handler(payload, stream)
+              : h.apiRoute.handler(stream)
+          );
+        }
+
+        const handler = async () => {
+          // handshake payload
+          const payload = req.payload();
+          return payload !== null
+            ? h.apiRoute.handler(payload, stream)
+            : h.apiRoute.handler(stream);
+        };
+
+        const mwRequest = new MiddlewareRequest(stream, undefined, undefined);
+        return invokeMiddlewareChain(mwRequest, middlewares, handler);
       }
     };
   }
 
-  if (h.raw) {
+  if (h.apiRoute.raw) {
     return {
-      ...h,
+      ...h.apiRoute,
       handler: (
         req: runtime.Request,
         resp: runtime.ResponseWriter,
         body: runtime.BodyReader
       ) => {
         setCurrentRequest(req);
+
         const rawReq = new RawRequest(req, body);
         const rawResp = new RawResponse(rawReq, resp);
-        return h.handler(rawReq, rawResp);
+
+        if (middlewares.length === 0) {
+          return toResponse(h.apiRoute.handler(rawReq, rawResp));
+        }
+
+        const handler = async () => {
+          return h.apiRoute.handler(rawReq, rawResp);
+        };
+
+        const mwRequest = new MiddlewareRequest(undefined, rawReq, rawResp);
+        return invokeMiddlewareChain(mwRequest, middlewares, handler);
       }
     };
   }
+
   return {
-    ...h,
+    ...h.apiRoute,
     handler: (req: runtime.Request) => {
       setCurrentRequest(req);
-      const payload = req.payload();
-      return payload !== null ? h.handler(payload) : h.handler();
+
+      if (middlewares.length === 0) {
+        const payload = req.payload();
+        return toResponse(
+          payload !== null ? h.apiRoute.handler(payload) : h.apiRoute.handler()
+        );
+      }
+
+      const handler = async () => {
+        const payload = req.payload();
+        return payload !== null
+          ? h.apiRoute.handler(payload)
+          : h.apiRoute.handler();
+      };
+
+      const mwRequest = new MiddlewareRequest(undefined, undefined, undefined);
+      return invokeMiddlewareChain(mwRequest, middlewares, handler);
     }
   };
 }
+
+function toResponse(payload: any): any {
+  if (payload instanceof Promise) {
+    return (async () => {
+      return new HandlerResponse(await payload);
+    })();
+  } else {
+    return new HandlerResponse(payload);
+  }
+}
```

Comment:
What's the purpose of this function?

Reply:
If the handler returns a promise it converts `Promise<any>` into `Promise<HandlerResponse{ payload: any }>`, otherwise the payload field would be a promise in the case that the handler is async